2/contrib/s2i/run

```diff
@@ -11,11 +11,117 @@ source /usr/local/bin/kube-slave-common.sh
+CONFIG_TIMESTAMP="${CONFIG_TIMESTAMP:-config-timestamp}"
+
+log() {
+  local level=$1; shift
+  local datetime=$(date "+%b %d, %Y %r")
+  echo "${datetime} ${level}: $*"
+}
+
+log_info() {
+  log INFO "$@"
+}
+
+log_warn() {
+  log WARN "$@"
+}
+
+config_changed() {
+  # don't worry about configurations if there isn't a dir
+  # but note that the image we are building already has the dir
+  [ -d ${IMAGE_CONFIG_DIR} ] || return 1
+
+  local img_config_timestamp="$IMAGE_CONFIG_DIR/$CONFIG_TIMESTAMP"
+
+  # treat absense of timestamp in configmap as timestamp changed to be on the
+  # safe side so that configuration files are generated
+  [ -f "$img_config_timestamp" ] || return 0
+
+  local our_timestamp="$JENKINS_HOME/$CONFIG_TIMESTAMP"
+  touch -a "$our_timestamp"
+
+  # return 0 if files are different
+  [[ $(cmp -s "$img_config_timestamp" "$our_timestamp") != 0 ]]
+}
+
+generate_jenkins_config()  {
+  local config_file="config.xml"
+  local config_tpl_path="$IMAGE_CONFIG_DIR/$config_file.tpl"
+
+  [ -f "$config_tpl_path" ] || {
+    log_warn "Skipping generation of '$config_file' as $config_tpl_path is not found"
+    return 0
+  }
+
+  # If it contains a template (tpl) file, we can do additional manipulations
+  # to customize the configuration.
+  export KUBERNETES_CONFIG=$(generate_kubernetes_config)
+
+  for name in $(get_is_names); do
+    log_info "... adding image ${name}:latest as Kubernetes slave"
+  done
+  log_info "Generating kubernetes-plugin configuration (${config_tpl_path})"
+  envsubst < "$config_tpl_path" > "${IMAGE_CONFIG_GEN_DIR}/$config_file"
+}
+
+generate_credentials()  {
+  local cred_file="credentials.xml"
+  local cred_tpl_path="$IMAGE_CONFIG_DIR/$cred_file.tpl"
+
+  [ -f "$cred_tpl_path" ] || {
+    log_warn "Skipping generation of '$cred_file' as $config_tpl_path is not found"
+    return 0
+  }
+
+  [ -z "${KUBERNETES_CONFIG}" ] && {
+    log_warn "Skipping generaton of kubernetes-plugin credentials as KUBERNETES_CREDENTIALS is blank"
+    return 0
+  }
+
+  log_info "Generating kubernetes-plugin credentials ($cred_file) ..."
+  export KUBERNETES_CREDENTIALS=$(generate_kubernetes_credentials)
+
+  # Fix the envsubst trying to substitute the $Hash inside credentials.xml
+  export Hash="\$Hash"
+  envsubst < "$cred_tpl_path" > "${IMAGE_CONFIG_GEN_DIR}/$cred_file"
+}
+
+copy_config_files() {
+  log_info "Copying Jenkins configuration to $JENKINS_HOME"
+  cp -frL "$IMAGE_CONFIG_DIR/"* "$JENKINS_HOME/"
+  cp -frL "$IMAGE_CONFIG_GEN_DIR/"* "$JENKINS_HOME/"
+
+  # we don't need the templates
+  log_info "Removing template files (*.tpl) from $JENKINS_HOME"
+  rm -f "$JENKINS_HOME/"*.tpl
+}
+
+
+generate_all_configs() {
+  mkdir -p "$IMAGE_CONFIG_GEN_DIR"
+  log_info "Using temp directory $IMAGE_CONFIG_GEN_DIR for generating configurations"
+  generate_jenkins_config
+  generate_credentials
+  copy_config_files
+  rm -fr "${IMAGE_CONFIG_GEN_DIR}"
```

Comment:
Same comment as before, I am not sure if it is really good idea to remove all created stuff,

Reply:
I am only removing the tmp directory used to expand the templates after they are copied to appropriate locations. So the workflow is like this ... 
1. check  if config needs to be generated
2. if so, create a tmp dir and generate all configs there
3. copy the expanded configs to appropriate location
4. cleanup the tmp dir i.e. rm it 

This is all done under $JENKINS_HOME which is required to by read-write mount

Reply:
Ok so this remove is just executed at most once, so when an instance is unidled, since everything is already in place then it is not executed right?

Reply:
@lordofthejars yes, and no. Given the current config-map no, the config will be generated every time the jenkins boots until we modify existing config-map to add a `timestamp` see - `CONFIG_TIMESTAMP` var above and how that is used `config_changed` method.

Reply:
well configs are not so many files IIRC and not so much MB so we are fine.